packages/classes/src/PayloadArray.js

```diff
@@ -0,0 +1,85 @@
+import { sortedIndexBy } from './tools/utils/sortedArray'
+import Payload from './Payload'
+
+/**
+ * PayloadArray stores a series of Payloads in a data structure that allows for easy filtering
+ * and lookup by position and type.
+ */
+class PayloadArray {
+	payloads = []
+
+	constructor(...payloads) {
+		payloads.forEach((payload) => this.insert(payload))
+	}
+
+	get length() { return this.payloads.length }
+
+	/**
+	 * Insert a Payload into the PayloadArray while maintaining position order.
+	 *
+	 * @param  {Payload} payload The Payload to be inserted.
+	 * @return {PayloadArray}    This PayloadArray to allow for chaining.
+	 */
+	insert = (payload) => {
+		if (!(payload instanceof Payload)) {
+			throw new Error('Attempt to insert a non-Payload to PayloadArray')
+		}
+		const index = this.indexOfPosition(payload.position)
+		this.payloads.splice(index, 0, payload)
+		return this
+	}
+
+	/**
+	 * Find the index to which a payload with a given position would be inserted.
+	 *
+	 * @param  {Number}  position      The position being searched for.
+	 * @param  {Boolean} returnHighest The position being searched for.
+	 * @return {Number}          The index of the item, or the PayloadArray length.
+	 */
+	indexOfPosition = (position, returnHighest = false) => sortedIndexBy(
+		this.payloads,
+		position,
+		(payload) => payload.position,
+		returnHighest,
+	)
+
+	/**
+	 * Create a new PayloadArray containing only Payloads of the specified type.
+	 *
+	 * @param  {String} type  The type of payload to include in the new PayloadArray.
+	 * @return {PayloadArray} A filtered PayloadArray.
+	 */
+	filterByType = (type) => {
+		const filteredPayloadArray = new PayloadArray()
+		const typeStem = type.endsWith('.ANY') ? type.slice(0, -3) : type
+		filteredPayloadArray.payloads = this.payloads.filter(
+			(payload) => payload.type.startsWith(typeStem),
+		)
+		return filteredPayloadArray
```

Comment:
It feels a little bit un-idiomatic to assign the payloads with `.payloads =` rather than passing it as the constructor argument to `new PayloadArray()`. Do you think it would be better to do something like:

```js
const filteredPayloads = this.payloads.filter(…)
return new PayloadArray(filteredPayloads)
```

I recognize this is slower, since you're having to use `insert` rather than just direct assignment, but it also feels safer somehow. Thoughts?

Reply:
My thinking was that that since this method is defined inside of the object it is OK to access object properties (if this was a method defined externally or in a utility I would be in 100% agreement).  I appreciate how that might not actually make it OK.

Here is the runtime cost analysis:

* Insert is an `O(log(n))` operation since it ultimately calls sortedIndexBy which is `O(log(n))`.
* Running insert for each of `n` items is then an `O(n * log(n))` operation.  This would mean that an `insert` based filter is `O(log(n))`
* Setting the array directly is constant time aka `O(1)`
* Setting the filtered array directly would make the filter whatever complexity it was (`O(n)` in the case of type filter, `O(log(n))` in the case of position filter)

So, using insert increases the runtime complexity somewhat significantly, and is undesirable in both cases from an efficiency perspective, though I appreciate that we want to ban direct access of `payloads` outside of this class.

Another option would be to create a `setInternalArray` method which would verify sort order (an O(n) operation).  This would at least be better than O(n*logn), but still not as fast as direct setting.  This would not impact the complexity of the type filter, but would impact the complexity of the position filter.

Reply:
The TL;DR is that I believe that since this is an internal method it's OK to access internal attributes directly that normally should not be accessed.  In this case, it buys us some potentially significant efficiency (which won't matter for short lists, but if we're dealing with say, 10,000 payloads it might)

Reply:
My experience with JS optimization is that expensive operations can have as much or more of an effect on runtime than algorithmic complexity. Paying some attention to big O generally is good, but also don't let the cost of c balloon or it'll eat the big O for breakfast. New-ing an object is expensive, setting a property is cheap, calling a method and copying an array are both middling costs. Premature optimizations that make systems more complicated and harder to read are usually a bad idea, optimizations that rely on simplifying the work involved without committing grave legibility sins are a good idea. So tl;dr ideologically I'm with @slifty on this.

Reply:
+1 to what Chris wrote, and I should take this moment to re-iterate that writing Data Structures in 2020 still feels like a fundamentally Bad Idea so I'm sorry :(

Reply:
This all makes a lot of sense, carry on!